### ISSUE TITLE
Add support for persistent map fds throughout multiple python calls

### ISSUE THREAD

**Issue by valkum** [iovisor/bcc#619]
I came across a situation where i want to have command line parameters for my python script for adding entries to a map or to remove entries from a map/clear a map.

For this purpose it would be nice if bcc would support some helper functions to store and load the map filedescriptors or attach the BPF object to an already running program.

I asked in irc and should open a feature request ticket. I haven't tried to pickle the whole bpf object, its on my todo list.

Maybe anyone has achieved what I try, if not such a feature would be very nice if you run your bpf program as a service rather than a utility.

**Comment by 4ast** [comment]
>  load the map filedescriptors or attach the BPF object to an already running program.

could you please describe your use case?

> to pickle the whole bpf object

what does it mean? :)

**Comment by valkum** [comment]
Lets assume i have a python script loading my bpf program. In there are is a map containing rules to drop a packet or not.
I don't want the python script to run the whole lifecycle of my bpf program.
The better solution would be to have my script and i can call ./script.py -a x to add x to my map.
For this i need to access the map of the running bpf program.


Whith pickly i meant the python object serialization (https://docs.python.org/2/library/pickle.html) to save the whole BPF object to disk between script calls.

**Comment by 4ast** [comment]
> For this i need to access the map of the running bpf program.

this can be achieved via bpffs.
the bpf program and maps can be pinned inside bpffs.
The initial script that loaded the program can exit and program (if it's networking)
can be still running.
Later on another script can start and get a map fd from bpffs.
Unfortunately bcc doesn't have support for bpffs yet. Only iproute2 does.
@borkmann

bpf objects belong to kernel so they cannot be serialized by python and any user space tools.

**Comment by valkum** [comment]
Yeah kinda thought serialization would not work.

I know about bpffs but did not know how the pinning exactly works. Found @borkmann's code from https://lkml.org/lkml/2015/10/29/270
I thought pinning is a must if you want to use them, so i though bcc would already pin the map fds.

So basically bcc needs to pin the map after creation. That seems quiet easy to implement on the first look into __init__.py and table.py. The question is if this is needed for every program, i don't think so, so somehow a parameter would be needed to tell get_table or the lib to pin the fd and to which path.

I'm not sure about the map access. Do we need to compile the BPF programm to tell the Table about our Map structure?

**Comment by 4ast** [comment]
yes. by default it shouldn't pin it, since Ctrl-C of .py process won't clean it up automatically.
sharing map layout between different processes is tricky.
could you please describe your use case in better detail? since probably only one-off solution is possible.

**Comment by valkum** [comment]
Right now im working with tc_cls/act.
So the following calls would be nice.
Calling:
script.py to load script.c
script.py -a <string> to add string to map relative to cache.c
script.py -r <string> to remove string from map.


My workaround for now is:
script.py -f <filename> to load script.c and add lines of filename into the map.
script.py -c to stop the bpf program completely by removing my tc act.
But this workaround is not really neat.

I can't get more into detail about the specifics but i think the calls above would clarify my use case.
